packages/agents/src/mcp/index.ts

```diff
@@ -82,36 +133,192 @@ export abstract class McpAgent<
   onStateUpdate(state: State | undefined, source: Connection | "server") {
     // override this to handle state updates
   }
+  async onStart() {
+    this.props = (await this.ctx.storage.get("props")) as Props;
+    this.init?.();
+
+    // Connect to the MCP server
+    this.#transport = new McpTransport(() => this.getWebSocket());
+    await this.server.connect(this.#transport);
+  }
 
   /**
    * McpAgent API
    */
   abstract server: McpServer;
-  private transport!: SSEEdgeServerTransport;
   props!: Props;
   initRun = false;
 
   abstract init(): Promise<void>;
 
   async _init(props: Props) {
+    this.ctx.storage.put("props", props);
     this.props = props;
     if (!this.initRun) {
       this.initRun = true;
       await this.init();
     }
   }
 
-  async onSSE(path: string): Promise<Response> {
-    this.transport = new SSEEdgeServerTransport(
-      `${path}/message`,
-      this.ctx.id.toString()
-    );
-    await this.server.connect(this.transport);
-    return this.transport.sseResponse;
+  async #initialize(): Promise<void> {
+    await this.ctx.blockConcurrencyWhile(async () => {
+      this.#status = "starting";
+      await this.onStart();
+      this.#status = "started";
+    });
   }
 
-  async onMCPMessage(request: Request): Promise<Response> {
-    return this.transport.handlePostMessage(request);
+  // Allow the worker to fetch a websocket connection to the agent
+  async fetch(request: Request): Promise<Response> {
+    if (this.#status !== "started") {
+      // This means the server "woke up" after hibernation
+      // so we need to hydrate it again
+      await this.#initialize();
+    }
+
+    // Only handle WebSocket upgrade requests
+    if (request.headers.get("Upgrade") !== "websocket") {
+      return new Response("Expected WebSocket Upgrade request", {
+        status: 400,
+      });
+    }
+
+    const url = new URL(request.url);
+    const sessionId = url.searchParams.get("sessionId");
+    if (!sessionId) {
+      return new Response("Missing sessionId", { status: 400 });
+    }
+
+    // Create a WebSocket pair
+    const webSocketPair = new WebSocketPair();
+    const [client, server] = Object.values(webSocketPair);
+
+    // Accept the WebSocket with hibernation support
+    this.ctx.acceptWebSocket(server);
```

Comment:
if we only care about the first websocket that's connected, do you want to throw an error if this is the second one?